File: avspec/tools/interview.mjs
#!/usr/bin/env node
// AVSpec guided-QA CLI — the deterministic (no-LLM) authoring driver.
// Same brain as the verifier (tools/lib/analyze.mjs); the CLI owns the
// mechanical side: show the ordered question queue, and apply structured
// answers as a patch to the manifest.
//
// Commands:
//   node tools/interview.mjs <dir> next [--json]
//       Print the ordered queue: blocking errors first, then the open
//       questions (todos) in authoring-layer order. Exit 0 if the spec passes
//       for its current status, else 1.
//
//   node tools/interview.mjs <dir> apply <patch.yaml> [--scaffold]
//       Merge a structured answer-patch into avspec.yaml (upserts requirements,
//       acceptance, components, contracts, decisions, tasks; sets tests; merges
//       layers). With --scaffold, also create any referenced-but-missing files
//       (artifact stubs, contract stubs, ADR stubs, .feature test stubs).
//       Re-analyzes and prints the todo delta.
//
//   node tools/interview.mjs <dir> ask
//       Interactive readline wizard: walk the todos, prompt for each, write the
//       answer, repeat. (Needs a TTY. The Claude skill is the richer
//       interactive option; this is the offline equivalent.)
//
// Patch shape (all sections optional; ids auto-assigned when omitted):
//   requirements: [{ id?, title, rationale?, acceptance: [{ id?, ears, test? }] }]
//   components:   [{ id?, responsibility, depends_on?, interfaces?: [{name, contract}] }]
//   contracts:    [{ id?, type, path }]
//   decisions:    [{ id?, title, status, path }]
//   tasks:        [{ id?, title, satisfies, touches?, depends_on?, parallelizable? }]
//   tests:        { AC-001: "verification/acceptance/REQ-001.feature#scenario" }
//   layers:       { presentation: [CMP-001], domain: [CMP-002], data: [CMP-003] }

import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import YAML from "js-yaml";
import { analyze, ordered } from "./lib/analyze.mjs";

const argv = process.argv.slice(2);
const flags = new Set(argv.filter((a) => a.startsWith("--")));
const pos = argv.filter((a) => !a.startsWith("--"));
const specDir = path.resolve(pos[0] || ".");
const cmd = pos[1] || "next";

const RULES_REL = "verification/architecture-rules.yaml";
const p = (rel) => path.join(specDir, rel);
const exists = (rel) => fs.existsSync(p(rel));

try {
  if (cmd === "next") cmdNext();
  else if (cmd === "apply") await cmdApply(pos[2]);
  else if (cmd === "ask") await cmdAsk();
  else { console.error(`unknown command: ${cmd}`); process.exit(2); }
} catch (e) {
  console.error(`error: ${e.message}`);
  process.exit(1);
}

// ---- commands ---------------------------------------------------------------
function cmdNext() {
  const r = requireAnalysis();
  const q = ordered(r.findings);
  const errs = q.filter((f) => f.severity === "error");
  const todos = q.filter((f) => f.severity === "todo");
  const warns = q.filter((f) => f.severity === "warn");

  if (flags.has("--json")) {
    console.log(JSON.stringify({ status: r.status, pass: r.pass, counts: r.counts,
      queue: [...errs, ...todos].map((f, i) => ({ n: i + 1, severity: f.severity, code: f.code, ids: f.ids, ask: f.question || f.message })) }, null, 2));
    process.exit(r.pass ? 0 : 1);
  }

  console.log(`\n${r.manifest.metadata.name} — status: ${r.status} — ${todos.length} open question(s), ${errs.length} blocking error(s)\n`);
  let n = 1;
  if (errs.length) {
    console.log("must fix first (spec is inconsistent):");
    for (const f of errs) console.log(`  ${n++}. [${f.code}] ${f.message}`);
    console.log("");
  }
  if (todos.length) {
    console.log("open questions:");
    for (const f of todos) console.log(`  ${n++}. ${f.question}`);
    console.log("");
  }
  if (!errs.length && !todos.length) console.log(`Nothing open. ${r.strict ? "Spec passes." : "Ready to set status: ready."}`);
  if (warns.length) console.log(`(${warns.length} warning(s) — non-blocking)`);
  process.exit(r.pass ? 0 : 1);
}

async function cmdApply(patchPath) {
  if (!patchPath) throw new Error("apply needs a patch file: interview.mjs <dir> apply <patch.yaml>");
  const before = requireAnalysis();
  const m = loadManifest();
  const patch = YAML.load(fs.readFileSync(path.resolve(patchPath), "utf8")) || {};

  const summary = applyPatch(m, patch);
  saveManifest(m);
  if (flags.has("--scaffold")) summary.scaffolded = scaffold(m);

  const after = analyze(specDir);
  console.log(`applied: ${summary.notes.join("; ") || "no manifest changes"}`);
  if (summary.scaffolded?.length) console.log(`scaffolded: ${summary.scaffolded.join(", ")}`);
  console.log(`todos: ${before.counts.todo} -> ${after.counts.todo}   errors: ${before.counts.error} -> ${after.counts.error}`);
  if (after.counts.todo === 0 && after.counts.error === 0 && !after.strict)
    console.log(`\nNo open todos. Set metadata.status: ready, then run: node tools/verify.mjs ${pos[0] || "."}`);
  process.exit(0);
}

async function cmdAsk() {
  const rl = (await import("node:readline/promises")).createInterface({ input: process.stdin, output: process.stdout });
  const ask = (q) => rl.question(q);
  try {
    for (;;) {
      const r = analyze(specDir);
      if (r.fatal) { console.log(r.fatal); break; }
      const q = ordered(r.findings);
      const err = q.find((f) => f.severity === "error");
      if (err) { console.log(`\nBlocking error — fix in the manifest, then rerun: [${err.code}] ${err.message}`); break; }
      const t = q.find((f) => f.severity === "todo");
      if (!t) { console.log(`\nAll questions answered (status: ${r.status}).`); if (!r.strict) console.log("Set metadata.status: ready to finish."); break; }

      const m = loadManifest();
      console.log(`\n${t.question}`);
      if (t.code === "NO_REQUIREMENTS" || t.code === "REQ_NO_AC") {
        const reqId = t.code === "NO_REQUIREMENTS" ? await newRequirement(m, ask) : t.ids[0];
        const ears = (await ask("  EARS criterion: ")).trim();
        if (ears) {
          const acId = nextId("AC-", allAcIds(m));
          const req = m.requirements.find((r2) => r2.id === reqId);
          const def = `verification/acceptance/${reqId}.feature#${acId.toLowerCase()}`;
          const test = (await ask(`  test ref [${def}]: `)).trim() || def;
          (req.acceptance ||= []).push({ id: acId, ears, test });
        }
      } else if (t.code === "AC_UNSATISFIED") {
        const title = (await ask("  task title: ")).trim();
        if (title) {
          const touches = (await ask("  touches (CMP-/CTR- ids, comma) []: ")).split(",").map((s) => s.trim()).filter(Boolean);
          (m.tasks ||= []).push({ id: nextId("TSK-", ids(m.tasks)), title, satisfies: [t.ids[0]], ...(touches.length ? { touches } : {}) });
        }
      } else if (t.code === "AC_NO_TEST") {
        const acId = t.ids[0];
        const ac = allAc(m).find((a) => a.id === acId);
        const def = `verification/acceptance/${ac.req}.feature#${acId.toLowerCase()}`;
        ac.entry.test = (await ask(`  test ref [${def}]: `)).trim() || def;
      } else if (t.code === "LAYER_UNPLACED") {
        const layer = (await ask("  layer (presentation/domain/data): ")).trim();
        if (layer) mergeLayers({ [layer]: [t.ids[0]] });
      } else if (t.code.endsWith("_MISSING")) {
        const yes = (await ask("  scaffold this file now? [Y/n]: ")).trim().toLowerCase();
        if (yes !== "n") { saveManifest(m); scaffold(loadManifest()); continue; }
      } else {
        console.log("  (no interactive handler; edit the manifest directly)"); break;
      }
      saveManifest(m);
    }
  } finally { rl.close(); }
}

// ---- patch application ------------------------------------------------------
function applyPatch(m, patch) {
  const notes = [];
  m.requirements ||= [];
  for (const pr of patch.requirements || []) {
    let req = pr.id && m.requirements.find((r) => r.id === pr.id);
    if (!req) { req = { id: pr.id || nextId("REQ-", ids(m.requirements)), title: pr.title || "(untitled)", ...(pr.rationale ? { rationale: pr.rationale } : {}), acceptance: [] }; m.requirements.push(req); notes.push(`+${req.id}`); }
    for (const pa of pr.acceptance || []) {
      const acId = pa.id || nextId("AC-", allAcIds(m));
      (req.acceptance ||= []).push({ id: acId, ears: pa.ears, ...(pa.test ? { test: pa.test } : {}) });
      notes.push(`+${acId}`);
    }
  }
  for (const [acId, ref] of Object.entries(patch.tests || {})) {
    const hit = allAc(m).find((a) => a.id === acId);
    if (hit) { hit.entry.test = ref; notes.push(`test:${acId}`); } else notes.push(`test:${acId}?missing`);
  }
  for (const pc of patch.components || []) upsert(m, "components", pc, "CMP-", notes);
  for (const pc of patch.contracts || []) upsert(m, "contracts", pc, "CTR-", notes);
  for (const pd of patch.decisions || []) upsert(m, "decisions", pd, "ADR-", notes);
  for (const pt of patch.tasks || []) upsert(m, "tasks", pt, "TSK-", notes);
  if (patch.layers) { mergeLayers(patch.layers); notes.push("layers"); }
  return { notes };
}

function upsert(m, key, entry, prefix, notes) {
  m[key] ||= [];
  const id = entry.id || nextId(prefix, ids(m[key]));
  const cur = m[key].find((x) => x.id === id);
  if (cur) { Object.assign(cur, entry, { id }); notes.push(`~${id}`); }
  else { m[key].push({ id, ...entry }); notes.push(`+${id}`); }
}

function mergeLayers(layers) {
  const rules = exists(RULES_REL) ? (YAML.load(fs.readFileSync(p(RULES_REL), "utf8")) || {}) : {};
  rules.layers ||= {};
  for (const [layer, comps] of Object.entries(layers)) {
    rules.layers[layer] = Array.from(new Set([...(rules.layers[layer] || []), ...comps]));
  }
  if (!rules.allow) rules.allow = [{ from: "presentation", to: ["domain"] }, { from: "domain", to: ["data"] }];
  fs.mkdirSync(path.dirname(p(RULES_REL)), { recursive: true });
  fs.writeFileSync(p(RULES_REL), YAML.dump(rules, { lineWidth: 120 }));
}

// ---- scaffolding ------------------------------------------------------------
function scaffold(m) {
  const made = [];
  const write = (rel, body) => { if (exists(rel)) return; fs.mkdirSync(path.dirname(p(rel)), { recursive: true }); fs.writeFileSync(p(rel), body); made.push(rel); };

  for (const [key, rel] of Object.entries(m.artifacts || {}))
    if (rel && !exists(rel)) write(rel, `# ${key[0].toUpperCase() + key.slice(1)}\n\n<!-- generated stub — expand the ${key} IDs from avspec.yaml here -->\n`);

  for (const c of m.contracts || [])
    if (!exists(c.path)) write(c.path, contractStub(c));

  for (const d of m.decisions || [])
    if (!exists(d.path)) write(d.path, `# ${d.id} — ${d.title}\n\n- **Status:** ${d.status}\n\n## Context\n\n<!-- stub -->\n\n## Decision\n\n## Consequences\n`);

  // group acceptance tests by feature file
  const byFile = new Map();
  for (const a of allAc(m)) {
    if (!a.entry.test) continue;
    const [file, scen] = a.entry.test.split("#");
    if (!file.endsWith(".feature") || !file.includes("/")) continue;
    (byFile.get(file) || byFile.set(file, []).get(file)).push({ ac: a.id, req: a.req, ears: a.entry.ears, scen: scen || a.id });
  }
  for (const [file, scens] of byFile) {
    if (exists(file)) continue;
    const req = scens[0].req;
    let body = `Feature: ${req}\n`;
    for (const s of scens) body += `\n  # ${s.ac}: ${s.ears || ""}\n  Scenario: ${s.scen}\n    Given <precondition>\n    When <action>\n    Then <observable outcome>\n`;
    write(file, body);
  }
  return made;
}

function contractStub(c) {
  if (c.type === "openapi") return `openapi: 3.1.0\ninfo: { title: ${path.basename(c.path)}, version: "0.0.0" }\npaths: {}\n`;
  if (c.type === "asyncapi") return `asyncapi: 3.0.0\ninfo: { title: ${path.basename(c.path)}, version: "0.0.0" }\nchannels: {}\n`;
  return `# JSON Schema stub (${c.id})\n$schema: "https://json-schema.org/draft/2020-12/schema"\ntype: object\n`;
}

// ---- manifest io + id helpers ----------------------------------------------
function loadManifest() { return YAML.load(fs.readFileSync(p("avspec.yaml"), "utf8")); }
function saveManifest(m) { fs.writeFileSync(p("avspec.yaml"), YAML.dump(m, { lineWidth: 120, noRefs: true })); }
function ids(arr) { return (arr || []).map((x) => x.id); }
function allAcIds(m) { return (m.requirements || []).flatMap((r) => (r.acceptance || []).map((a) => a.id)); }
function allAc(m) { return (m.requirements || []).flatMap((r) => (r.acceptance || []).map((a) => ({ id: a.id, req: r.id, entry: a }))); }

function nextId(prefix, existing) {
  const nums = existing.filter((s) => s.startsWith(prefix)).map((s) => parseInt(s.slice(prefix.length), 10)).filter((n) => !Number.isNaN(n));
  const width = Math.max(3, ...existing.filter((s) => s.startsWith(prefix)).map((s) => s.slice(prefix.length).length));
  const n = (nums.length ? Math.max(...nums) : 0) + 1;
  return prefix + String(n).padStart(width, "0");
}

async function newRequirement(m, ask) {
  const title = (await ask("  requirement title: ")).trim() || "(untitled)";
  const id = nextId("REQ-", ids(m.requirements));
  (m.requirements ||= []).push({ id, title, acceptance: [] });
  return id;
}

function requireAnalysis() {
  const r = analyze(specDir);
  if (r.fatal) { console.error(`FATAL: ${r.fatal}`); process.exit(1); }
  return r;
}

File: avspec/src/avspec/interview.py
"""Module-first AVSpec interview."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import typer
from rich.console import Console

from avspec.analysis import analyze
from avspec.findings import Finding, ordered
from avspec.io import load_manifest, save_raw_manifest, yaml

console = Console()


def ask(spec_dir: Path) -> int:
    """Run a module-first interview until the next manual stop point."""
    resolved = spec_dir.resolve()
    while True:
        report = analyze(resolved)
        if report.fatal:
            console.print(f"FATAL: {report.fatal}", style="red")
            return 1

        findings = ordered(report.findings)
        error = next((finding for finding in findings if finding.severity == "error"), None)
        if error:
            console.print(
                f"\nBlocking error - fix the manifest, then rerun: [{error.code}] {error.message}"
            )
            return 1

        todo = next((finding for finding in findings if finding.severity == "todo"), None)
        if todo is None:
            console.print(f"\nAll questions answered (status: {report.status}).")
            if not report.strict:
                console.print("Set metadata.status: ready after reviewing the generated artifacts.")
            return 0

        loaded = load_manifest(resolved)
        if loaded.raw is None:
            console.print("No manifest data loaded.", style="red")
            return 1

        changed = _handle_todo(resolved, loaded.raw, todo)
        if not changed:
            console.print("No change made; stopping so the interview does not loop.")
            return 1
        save_raw_manifest(resolved, loaded.raw)
        sync_artifacts(resolved, loaded.raw)


def _handle_todo(spec_dir: Path, manifest: dict[str, Any], finding: Finding) -> bool:
    console.print(f"\n[{finding.code}] {finding.question or finding.message}", style="bold")
    match finding.code:
        case "NO_COMPONENTS":
            return _ask_modules(manifest)
        case "NO_BOUNDARIES":
            return _ask_boundaries(spec_dir, manifest)
        case "NO_CONTRACTS" | "CMP_NO_CONTRACT":
            return _ask_contracts(manifest, finding.ids[0] if finding.ids else None)
        case "NO_CONSTITUTION":
            return _ask_constitution(manifest)
        case "NO_REQUIREMENTS":
            return _ask_capabilities(manifest)
        case "REQ_NO_AC":
            return _ask_acceptance(manifest, finding.ids[0])
        case "AC_UNSATISFIED":
            return _ask_task(manifest, finding.ids[0])
        case "AC_NO_TEST":
            return _ask_test(manifest, finding.ids[0])
        case "CMP_NO_STACK":
            return _ask_module_stack(manifest, finding.ids[0])
        case "NO_STACK" | "STACK_INCOMPLETE":
            return _ask_stack(manifest)
        case "LAYER_UNPLACED":
            return _ask_layer(spec_dir, finding.ids[0])
        case code if code.endswith("_MISSING"):
            return _scaffold_missing(spec_dir, manifest, finding)
        case _:
            console.print(
                "No interactive handler for this finding yet; edit the manifest directly."
            )
            return False


def _ask_stack(manifest: dict[str, Any]) -> bool:
    stack = _prompt_stack("suite default")
    if stack is None:
        return False
    manifest["stack"] = stack
    manifest["avspec"] = "0.2"
    return True


def _ask_module_stack(manifest: dict[str, Any], component_id: str) -> bool:
    component = next(
        (item for item in manifest.get("components") or [] if item["id"] == component_id),
        None,
    )
    if component is None:
        return False
    stack = _prompt_stack(f"{component_id} {_component_label(component)}")
    if stack is None:
        return False
    component["stack"] = stack
    manifest["avspec"] = "0.2"
    return True


def _prompt_stack(label: str) -> dict[str, Any] | None:
    console.print(f"Implementation stack for {label}.")
    languages = _csv_prompt("Implementation languages, comma-separated")
    if not languages:
        return None

    stack: dict[str, Any] = {"languages": languages}
    package_manager = typer.prompt(
        "Package manager or build tool", default="", show_default=False
    ).strip()
    frameworks = _csv_prompt("Frameworks/runtime libraries, comma-separated", required=False)
    bdd = typer.prompt("BDD/test style, if known", default="", show_default=False).strip()

    commands: dict[str, str] = {}
    for key in ("install", "lint", "typecheck", "test", "arch"):
        value = typer.prompt(f"Command for {key}, if known", default="", show_default=False).strip()
        if value:
            commands[key] = value

    if package_manager:
        stack["package_manager"] = package_manager
    if frameworks:
        stack["frameworks"] = frameworks
    if bdd:
        stack["bdd"] = bdd
    if commands:
        stack["commands"] = commands

    return stack


def _ask_modules(manifest: dict[str, Any]) -> bool:
    console.print("Define modules first. Use apps, services, shared packages, or integrations.")
    components = _ensure_list(manifest, "components")
    before = len(components)

    while True:
        name = typer.prompt("Module name (blank when done)", default="", show_default=False).strip()
        if not name:
            break
        responsibility = typer.prompt("What is this module responsible for?").strip()
        if not responsibility:
            continue
        component_id = next_id("CMP-", [item["id"] for item in components])
        components.append({"id": component_id, "responsibility": f"{name} - {responsibility}"})
        console.print(f"  added {component_id}: {name}")

    if len(components) == before:
        return False

    _ask_dependencies(components)
    return True


def _ask_boundaries(spec_dir: Path, manifest: dict[str, Any]) -> bool:
    components = manifest.get("components") or []
    if not components:
        return False

    console.print("Assign each module to an architecture layer.")
    layers: dict[str, list[str]] = {}
    for component in components:
        group = typer.prompt(
            f"Architecture layer for {component['id']} {_component_label(component)}",
            default="application",
            show_default=True,
        ).strip()
        layers.setdefault(group or "application", []).append(component["id"])

    allow: list[dict[str, Any]] = []
    groups = list(layers)
    console.print("Architecture layers: " + ", ".join(groups))
    for group in groups:
        targets = _csv_prompt(f"{group} may depend on which layers?", required=False)
        if targets:
            allow.append({"from": group, "to": targets})

    _write_rules(spec_dir, {"layers": layers, "allow": allow})
    return True


def _ask_contracts(manifest: dict[str, Any], component_id: str | None) -> bool:
    components = manifest.get("components") or []
    if not components:
        return False
    contracts = _ensure_list(manifest, "contracts")
    before = len(contracts)

    targets = [
        component
        for component in components
        if component_id is None or component["id"] == component_id
    ]
    for component in targets:
        console.print(f"\nModule boundary: {component['id']} - {_component_label(component)}")
        exposes = typer.confirm("Does this module expose an API or message contract?", default=True)
        if not exposes:
            component.setdefault("interfaces", [])
            continue
        interface_name = typer.prompt(
            "Interface name", default=_default_interface_name(component), show_default=True
        ).strip()
        contract_type = typer.prompt("Contract type", default="openapi", show_default=True).strip()
        contract_id = next_id("CTR-", [item["id"] for item in contracts])
        contract_path = typer.prompt(
            "Contract path",
            default=f"contracts/{interface_name.lower().replace(' ', '-')}.openapi.yaml",
            show_default=True,
        ).strip()
        contracts.append({"id": contract_id, "type": contract_type, "path": contract_path})
        component.setdefault("interfaces", []).append(
            {"name": interface_name, "contract": contract_id}
        )

    return len(contracts) > before


def _ask_dependencies(components: list[dict[str, Any]]) -> None:
    console.print("\nDeclare module dependencies by ID. Leave blank when none.")
    valid_ids = {component["id"] for component in components}
    for component in components:
        label = _component_label(component)
        deps = _csv_prompt(f"{component['id']} {label} depends on", required=False)
        selected = [dep for dep in deps if dep in valid_ids and dep != component["id"]]
        if selected:
            component["depends_on"] = selected


def _ask_constitution(manifest: dict[str, Any]) -> bool:
    console.print(
        "Capture the suite constitution before capabilities. "
        "These are top-level principles or constraints that every module must honor."
    )
    principles = _ensure_list(manifest, "principles")
    constraints = _ensure_list(manifest, "constraints")
    before = len(principles) + len(constraints)

    while True:
        statement = typer.prompt(
            "Principle or constraint (blank when done)", default="", show_default=False
        ).strip()
        if not statement:
            break
        kind = (
            typer.prompt("Type: principle or constraint", default="principle", show_default=True)
            .strip()
            .lower()
        )
        if kind.startswith("c"):
            constraints.append(
                {
                    "id": next_id("CON-", [item["id"] for item in constraints]),
                    "statement": statement,
                }
            )
        else:
            principles.append(
                {"id": next_id("PR-", [item["id"] for item in principles]), "statement": statement}
            )

    return len(principles) + len(constraints) > before


def _ask_capabilities(manifest: dict[str, Any]) -> bool:
    components = manifest.get("components") or []
    if components:
        console.print(
            "Now define suite capabilities. These are still high-level; no acceptance tests yet."
        )
        for component in components:
            console.print(f"  {component['id']} - {_component_label(component)}")
    requirements = _ensure_list(manifest, "requirements")
    before = len(requirements)

    while True:
        title = typer.prompt(
            "Capability title (blank when done)", default="", show_default=False
        ).strip()
        if not title:
            break
        rationale = typer.prompt(
            "Why does this capability matter?", default="", show_default=False
        ).strip()
        requirement: dict[str, Any] = {
            "id": next_id("REQ-", [item["id"] for item in requirements]),
            "title": title,
            "acceptance": [],
        }
        if rationale:
            requirement["rationale"] = rationale
        requirements.append(requirement)

    return len(requirements) > before


def _ask_acceptance(manifest: dict[str, Any], requirement_id: str) -> bool:
    requirement = _find_requirement(manifest, requirement_id)
    if requirement is None:
        return False
    console.print(
        "Acceptance comes after modules and capabilities. "
        f"Requirement: {requirement_id} - {requirement['title']}"
    )
    acceptance = requirement.setdefault("acceptance", [])
    before = len(acceptance)

    while True:
        ears = typer.prompt(
            "Observable criterion in EARS form (blank when done)", default="", show_default=False
        ).strip()
        if not ears:
            break
        ac_id = next_id("AC-", all_ac_ids(manifest))
        acceptance.append({"id": ac_id, "ears": ears})
        console.print(f"  added {ac_id}")

    return len(acceptance) > before


def _ask_task(manifest: dict[str, Any], ac_id: str) -> bool:
    title = typer.prompt(f"Task that delivers {ac_id}", default="", show_default=False).strip()
    if not title:
        return False
    touches = _csv_prompt("Touched module/component/contract IDs, comma-separated", required=False)
    task: dict[str, Any] = {
        "id": next_id("TSK-", [item["id"] for item in manifest.get("tasks") or []]),
        "title": title,
        "satisfies": [ac_id],
    }
    if touches:
        task["touches"] = touches
    _ensure_list(manifest, "tasks").append(task)
    return True


def _ask_test(manifest: dict[str, Any], ac_id: str) -> bool:
    hit = _find_acceptance(manifest, ac_id)
    if hit is None:
        return False
    requirement_id, acceptance = hit
    default = f"verification/acceptance/{requirement_id}.feature#{ac_id.lower()}"
    test_ref = typer.prompt("Executable test reference", default=default, show_default=True).strip()
    acceptance["test"] = test_ref
    return True


def _ask_layer(spec_dir: Path, component_id: str) -> bool:
    layer = typer.prompt("Layer", default="application", show_default=True).strip()
    if not layer:
        return False
    _merge_layers(spec_dir, {layer: [component_id]})
    return True


def _scaffold_missing(spec_dir: Path, manifest: dict[str, Any], finding: Finding) -> bool:
    if not typer.confirm("Scaffold missing referenced files now?", default=True):
        return False
    made = scaffold(spec_dir, manifest)
    if made:
        console.print("scaffolded: " + ", ".join(made))
    return bool(made)


def scaffold(spec_dir: Path, manifest: dict[str, Any]) -> list[str]:
    made: list[str] = []

    def write(rel: str, body: str) -> None:
        path = spec_dir / rel
        if path.exists():
            return
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(body)
        made.append(rel)

    for contract in manifest.get("contracts") or []:
        rel = contract.get("path")
        if not rel:
            continue
        if contract.get("type") == "openapi":
            write(
                rel,
                "openapi: 3.1.0\n"
                f'info: {{ title: {Path(rel).name}, version: "0.0.0" }}\n'
                "paths: {}\n",
            )
        elif contract.get("type") == "asyncapi":
            write(
                rel,
                "asyncapi: 3.0.0\n"
                f'info: {{ title: {Path(rel).name}, version: "0.0.0" }}\n'
                "channels: {}\n",
            )
        else:
            write(
                rel,
                "# AVSPEC:STUB\n"
                '$schema: "https://json-schema.org/draft/2020-12/schema"\n'
                "type: object\n",
            )

    for requirement in manifest.get("requirements") or []:
        for acceptance in requirement.get("acceptance") or []:
            test_ref = acceptance.get("test")
            if not test_ref or "/" not in test_ref:
                continue
            rel = test_ref.split("#", maxsplit=1)[0].split("::", maxsplit=1)[0]
            if not rel.endswith(".feature"):
                continue
            write(
                rel,
                f"Feature: {requirement['id']}\n\n"
                f"  # {acceptance['id']}: {acceptance.get('ears', '')}\n"
                f"  Scenario: {test_ref.split('#', maxsplit=1)[-1]}\n"
                "    Given <precondition>\n"
                "    When <action>\n"
                "    Then <observable outcome>\n",
            )

    return made


def sync_artifacts(spec_dir: Path, manifest: dict[str, Any]) -> None:
    artifacts = manifest.get("artifacts") or {}
    if artifacts.get("constitution"):
        (spec_dir / artifacts["constitution"]).write_text(_constitution_markdown(manifest))
    if artifacts.get("requirements"):
        (spec_dir / artifacts["requirements"]).write_text(_requirements_markdown(manifest))
    if artifacts.get("design"):
        (spec_dir / artifacts["design"]).write_text(_design_markdown(manifest))
    if artifacts.get("tasks"):
        (spec_dir / artifacts["tasks"]).write_text(_tasks_markdown(manifest))


def _constitution_markdown(manifest: dict[str, Any]) -> str:
    lines = ["# Constitution - " + manifest["metadata"]["name"], ""]
    lines.append("## Principles")
    principles = manifest.get("principles") or []
    lines.extend(
        [f"- **{item['id']}** - {item['statement']}" for item in principles]
        or ["No principles captured yet."]
    )
    lines.extend(["", "## Constraints"])
    constraints = manifest.get("constraints") or []
    lines.extend(
        [f"- **{item['id']}** - {item['statement']}" for item in constraints]
        or ["No constraints captured yet."]
    )
    return "\n".join(lines) + "\n"


def _requirements_markdown(manifest: dict[str, Any]) -> str:
    lines = ["# Requirements - " + manifest["metadata"]["name"], ""]
    requirements = manifest.get("requirements") or []
    if not requirements:
        lines.append("No capabilities captured yet.")
        return "\n".join(lines) + "\n"
    for requirement in requirements:
        lines.extend([f"## {requirement['id']} - {requirement['title']}", ""])
        if requirement.get("rationale"):
            lines.extend([f"**Why:** {requirement['rationale']}", ""])
        acceptance = requirement.get("acceptance") or []
        if not acceptance:
            lines.extend(["Acceptance criteria not captured yet.", ""])
            continue
        for item in acceptance:
            lines.append(f"- **{item['id']}** - {item['ears']}")
            if item.get("test"):
                lines.append(f"  Test: `{item['test']}`")
        lines.append("")
    return "\n".join(lines)


def _design_markdown(manifest: dict[str, Any]) -> str:
    lines = ["# Design - " + manifest["metadata"]["name"], ""]
    description = manifest["metadata"].get("description")
    if description:
        lines.extend(["## Context", "", description, ""])
    lines.extend(["## Modules", ""])
    components = manifest.get("components") or []
    if not components:
        lines.append("No modules captured yet.")
        return "\n".join(lines) + "\n"
    for component in components:
        lines.append(f"- **{component['id']}** - {_component_label(component)}")
        deps = ", ".join(component.get("depends_on") or []) or "none"
        lines.append(f"  - Depends on: {deps}")
        if component.get("stack"):
            languages = ", ".join(component["stack"].get("languages") or [])
            lines.append(f"  - Stack: {languages or 'not captured'}")
    lines.append("")
    return "\n".join(lines)


def _tasks_markdown(manifest: dict[str, Any]) -> str:
    lines = ["# Tasks - " + manifest["metadata"]["name"], ""]
    tasks = manifest.get("tasks") or []
    if not tasks:
        lines.append("No tasks captured yet.")
        return "\n".join(lines) + "\n"
    for task in tasks:
        lines.append(f"- **{task['id']}** - {task['title']}")
        lines.append(f"  Satisfies: {', '.join(task.get('satisfies') or [])}")
        if task.get("touches"):
            lines.append(f"  Touches: {', '.join(task['touches'])}")
    return "\n".join(lines) + "\n"


def _merge_layers(spec_dir: Path, layers: dict[str, list[str]]) -> None:
    rules_path = spec_dir / "verification/architecture-rules.yaml"
    rules_path.parent.mkdir(parents=True, exist_ok=True)
    rules = yaml.load(rules_path.read_text()) if rules_path.exists() else {}
    if not isinstance(rules, dict):
        rules = {}
    rules.setdefault("layers", {})
    for layer, component_ids in layers.items():
        current = rules["layers"].setdefault(layer, [])
        for component_id in component_ids:
            if component_id not in current:
                current.append(component_id)
    rules.setdefault(
        "allow",
        [
            {"from": "application", "to": ["platform", "integration"]},
            {"from": "platform", "to": ["integration"]},
        ],
    )
    with rules_path.open("w") as handle:
        yaml.dump(rules, handle)


def _write_rules(spec_dir: Path, rules: dict[str, Any]) -> None:
    rules_path = spec_dir / "verification/architecture-rules.yaml"
    rules_path.parent.mkdir(parents=True, exist_ok=True)
    with rules_path.open("w") as handle:
        yaml.dump(rules, handle)


def _ensure_list(manifest: dict[str, Any], key: str) -> list[dict[str, Any]]:
    value = manifest.setdefault(key, [])
    if not isinstance(value, list):
        raise TypeError(f"{key} must be a list")
    return value


def _csv_prompt(prompt: str, *, required: bool = True) -> list[str]:
    while True:
        raw = typer.prompt(prompt, default="", show_default=False).strip()
        values = [item.strip() for item in raw.split(",") if item.strip()]
        if values or not required:
            return values


def next_id(prefix: str, existing: list[str]) -> str:
    numbers = [
        int(item.removeprefix(prefix))
        for item in existing
        if item.startswith(prefix) and item.removeprefix(prefix).isdigit()
    ]
    width = max(
        [3, *[len(item.removeprefix(prefix)) for item in existing if item.startswith(prefix)]]
    )
    return prefix + str((max(numbers) if numbers else 0) + 1).zfill(width)


def all_ac_ids(manifest: dict[str, Any]) -> list[str]:
    return [
        acceptance["id"]
        for requirement in manifest.get("requirements") or []
        for acceptance in requirement.get("acceptance") or []
    ]


def _find_requirement(manifest: dict[str, Any], requirement_id: str) -> dict[str, Any] | None:
    return next(
        (item for item in manifest.get("requirements") or [] if item["id"] == requirement_id), None
    )


def _find_acceptance(manifest: dict[str, Any], ac_id: str) -> tuple[str, dict[str, Any]] | None:
    for requirement in manifest.get("requirements") or []:
        for acceptance in requirement.get("acceptance") or []:
            if acceptance["id"] == ac_id:
                return requirement["id"], acceptance
    return None


def _component_label(component: dict[str, Any]) -> str:
    return str(component.get("responsibility") or "(no responsibility)")


def _default_interface_name(component: dict[str, Any]) -> str:
    raw = _component_label(component).split("-", maxsplit=1)[0].strip()
    words = [word for word in raw.replace("_", " ").split(" ") if word]
    if not words:
        return f"{component['id']}API"
    return "".join(word[:1].upper() + word[1:] for word in words) + "API"
